sillsdev/languageforge-lexbox
# configure openid connect

this PR enables the use of openID connect to authenticate 3rd party apps with lexbox. For now this is disabled except in local dev, it's pending key storage decisions. We need to store signing and encryption keys somewhere and I'm not sure what the best option is right now for production.

How to test:
1. open this page https://oidcdebugger.com/
3. enter the following
4. authorize URL: `http://localhost:3000/api/login/open-id-auth`
5. clinet id: `becf2856-0690-434b-b192-a4032b72067f`
6. scope: `openid profile`
7. response type, check only code
8. token URI: http://localhost:3000/api/login/token
9. response mode query
10. click Orange send request button
11. if you're not logged in go to the next step, otherwise go to step 13
12.  you should land at the lexbox login page with a large `ReturnUrl` query parameter in the URL, login like normal (google login unknown?), once login is finished proceed to the next step
13. if you haven't approved this app before go to the next step otherwise go to step 15
14. you should land on a page asking if you want to permit "Oidc Debugger" to access your data: 
![image](https://github.com/sillsdev/languageforge-lexbox/assets/4575355/825b1f64-a0ce-46db-8a69-c67837b18f8f)
15. Click Approve to let the debugger have access
16. should arrive at a page like this:
![image](https://github.com/sillsdev/languageforge-lexbox/assets/4575355/601cdf40-49ad-4b14-917d-93a361155bc9)
17. scroll down and you should see a section labeled ID Token, you should be able to find your name in the payload json object.

if you want to test out the access token (first box, very large token), you can execute a request like this:
```
GET https://localhost:3000/api/AuthTesting/requires-auth
Authorization: Bearer {token here}
```

it should return success with your users name. This represents a token that could be used by a 3rd party app like Language Forge, FLEx lite, or Living Dictionaries to access APIs on behalf of the user.

flow diagram:
```mermaid
flowchart TD
    approve{Has approved already}
    approveUser(Do you approve?)
    start[debugger authorize] -->|Logged in| approve
    start -->|Not logged in| login(Login)
    login --> approve
    approve -->|no| approveUser
    approve -->|yes| debugger
    approveUser -->|"Approve (success)"| debugger
    approveUser -->|"Deny (fail)"| debugger
```
